README
NAME
    Lingua::ZH::Romanize::Pinyin - Romanization of Standard Chinese language

SYNOPSIS
        use Lingua::ZH::Romanize::Pinyin;

        my $conv = Lingua::ZH::Romanize::Pinyin->new();
        my $roman = $conv->char( $hanji );
        printf( "<ruby><rb>%s</rb><rt>%s</rt></ruby>", $hanji, $roman );

        my @array = $conv->string( $string );
        foreach my $pair ( @array ) {
            my( $raw, $ruby ) = @$pair;
            if ( defined $ruby ) {
                printf( "<ruby><rb>%s</rb><rt>%s</rt></ruby>", $raw, $ruby );
            } else {
                print $raw;
            }
        }

DESCRIPTION
    Pinyin is a phonemic notation of Chinese characters.

  $conv = Lingua::ZH::Romanize::Pinyin->new();
    This constructer methods returns a new object with its dictionary
    cached.

  $roman = $conv->char( $hanji );
    This method returns romanized letters of a Hanji character. It returns
    undef when $hanji is not a valid Hanji character. The argument's
    encoding must be UTF-8. Both of Simplified Chinese and Traditional
    Chinese are allowed.

  $roman = $conv->chars( $string );
    This method returns romanized letters of Hanji characters.

  @array = $conv->string( $string );
    This method returns a array of referenced arrays which are pairs of a
    Hanji chacater and its romanized letters.

        $array[0]           # first Chinese character's pair (array)
        $array[1][0]        # secound Chinese character itself
        $array[1][1]        # its romanized letters

DICTIONARY
    This module's Hanji to roman mapping table is based on PY.tit file which
    is distributed with cxterm.

MODULE DEPENDENCIES
    Storable module is required.

SEE ALSO
    Lingua::ZH::Romanize::Cantonese Lingua::JA::Romanize::Japanese
    Lingua::KO::Romanize::Hangul

    http://www.kawa.net/works/perl/romanize/romanize-e.html

AUTHOR
    Yusuke Kawasaki http://www.kawa.net/

COPYRIGHT
    Copyright (c) 2003-2006 Yusuke Kawasaki. All rights reserved.

LICENSE
    Any commercial use of the Software requires a license directly from the
    author(s). Please contact the author(s) to negotiate an appropriate
    license. Commercial use includes integration of all or part of the
    binary or source code covered by this permission notices into a product
    for sale or license to third parties on your behalf, or distribution of
    the binary or source code to third parties that need it to utilize a
    product sold or licensed on your behalf.



Changes
# Lingua::ZH::Romanize::* modules Changes

2006/04/04 (0.12)
	- cxterm/dict/gb/CCDOSPY.tit removed from archive which is not used
	- chars() method added

2006/03/26 (0.11)
	- first release


lib/Lingua/ZH/Romanize/DictZH.pm
=head1 NAME

Lingua::ZH::Romanize::DictZH - Dictionary converter

=head1 SYNOPSIS

    perl -MLingua::ZH::Romanize::DictZH -e 'Lingua::ZH::Romanize::DictZH->update();'

=head1 DESCRIPTION

This module creates dictionary cache files for
L<Lingua::ZH::Romanize::Pinyin> module.

Source dictionary files: (included in this package)

    Lingua/ZH/Romanize/big5/CTLauBig5.tit
    Lingua/ZH/Romanize/big5/PY.tit
    Lingua/ZH/Romanize/gb/CTLau.tit
    Lingua/ZH/Romanize/gb/PY.tit

Cached dictionary files:

    Lingua/ZH/Romanize/Cantonese.store
    Lingua/ZH/Romanize/Pinyin.store

DictZH is called only from Makefile.PL in the Pinyin package.

=head1 REQUIRED MODULES

L<Storable> module is required to create cached dictionary files.
Both of L<Unicode::Map> and L<Unicode::String> modules are
required on Perl 5.005 and 5.6.x to install the Pinyin package.
L<Encode> module is used on Perl 5.8.x.

=head1 SEE ALSO

L<Lingua::ZH::Romanize::Pinyin>

=cut
# ----------------------------------------------------------------
    package Lingua::ZH::Romanize::DictZH;
    use strict;
    use vars qw( $VERSION );
    $VERSION = "0.11";
# ----------------------------------------------------------------
    my $USE_ENCODE_PM = ( $] >= 5.008 );    # Perl 5.8.x or not
#   $USE_ENCODE_PM = 1;     # force to use Encode.pm
#   $USE_ENCODE_PM = 0;     # force to use Unicode::Map and Unicode::String
# ----------------------------------------------------------------
#   my $SKIP_IF_EXISTS = 1; # don't update when .store file exists
    my $SKIP_IF_EXISTS = 0; # override .store file when it exists
# ----------------------------------------------------------------
    my $DICT_FILES = {
        "Cantonese.store"   =>  [qw(
			cxterm/dict/big5/CTLauBig5.tit
			cxterm/dict/gb/CTLau.tit
		)],
        "Pinyin.store"      =>  [qw(
			cxterm/dict/big5/PY.tit
			cxterm/dict/gb/PY.tit
		)],
    };
# ----------------------------------------------------------------
sub update {
    my $package = shift;
    my $base = shift;
    unless ( defined $base ) {
        $base = $INC{join( "/", split("::",(__PACKAGE__).".pm"))};
        $base =~ s#/[^/]*$##;
    }

    my $update = 0;
    foreach my $storename ( keys %$DICT_FILES ) {
        my $storepath = $base."/".$storename;
        if ( -r $storepath ) {
            warn "Already-Exist: $storepath\n";
        } else {
            $update ++;
        }
    }
    return if ( $SKIP_IF_EXISTS && ! $update );     # no need to update

    warn "Loading-Module: Storable.pm\n";
    &require_storable();                # required

    if ( $USE_ENCODE_PM ) {
        warn "Loading-Module: Encode.pm\n";
        &require_encode();              # Perl 5.8.x
    } else {
        warn "Loading-Module: Unicode::Map\n";
        &require_unicode_map();         # Perl 5.005/5.6.x
        warn "Loading-Module: Unicode::String\n";
        &require_unicode_string();      #
    }

    foreach my $storename ( keys %$DICT_FILES ) {
        my $hash = {};
        foreach my $titpath ( @{$DICT_FILES->{$storename}} ) {
            warn "Loading-Dictionary: $titpath\n";
            $hash = &read_tit_dict( $titpath, $hash );
        }
        foreach my $key ( keys %$hash ) {
            next unless ref $hash->{$key};
            my $uniq = { map {$_=>1} @{$hash->{$key}} };
            $hash->{$key} = join( "/", sort keys %$uniq );
        }
        my $storepath = $base."/".$storename;
        warn "Writing-Storable: $storepath\n";
        Storable::store( $hash, $storepath ) or die "$! - $storename\n";
    }

    undef;
}
# ----------------------------------------------------------------
sub read_tit_dict {
    my $titname = shift or return;
    my $hash = shift || {};
    my $cmap = {qw( GB GB2312 BIG5 BIG5 KS EUC-KR JIS EUC-JP )};
    my $unistr = Unicode::String->new() unless $USE_ENCODE_PM;

    # find ENCODE: and wait until BEGINDICTIONARY
    open( TIT, $titname ) or die "$! - $titname\n";
    my $code;
    while( <TIT> ) {
        next if /^#/;
        $code = $cmap->{uc($1)} if ( /^ENCODE:\s*(\S+)/ );
        last if /^BEGINDICTIONARY/;
    }
    warn "Dictionary-Encoding: $code\n" if $code;
    my $unimap = Unicode::Map->new( $code ) unless $USE_ENCODE_PM;

    while( my $line = <TIT> ) {
        next if ( $line =~ /^#/ );
        chomp $line;

        my( $roman, $kanji ) = split( /\s+/, $line, 2 );
        $roman =~ s/^\\[0-7]{3}//s;
        $roman =~ s/\d+$//s;

        # convert encoding from GB/BIG5 to UTF-8
        if ( $code ) {
            if ( $USE_ENCODE_PM ) {
                Encode::from_to( $kanji, $code, "UTF-8" );   # GB/BIG5 to UTF-8
            } else {
                my $utf16 = $unimap->to_unicode( $kanji );   # GB/BIG5 to UCS2
                $unistr->ucs2( $utf16 );
                $kanji = $unistr->utf8();                    # UCS2 to UTF-8
            }
        }

        # split every UTF-8 wide characters
        while( $kanji =~ /([\300-\377][\200-\277]+)/g ) {
            my $char = $1;
            if ( ! exists $hash->{$char} ) {
                $hash->{$char} = $roman;
            } elsif ( ! ref $hash->{$char} ) {
                $hash->{$char} = [ $hash->{$char}, $roman ];
            } else {
                push( @{$hash->{$char}}, $roman );
            }
        }
    }
    close( TIT );

    $hash;
}
# ----------------------------------------------------------------
sub require_storable {
    return if defined $Storable::VERSION;
    local $@;
    eval { require Storable; };
    die "Storable module is required.\n" if $@;
}
# ----------------------------------------------------------------
sub require_encode {
    return if defined $Encode::VERSION;
    local $@;
    eval { require Encode; };
    die "Encode module is required.\n" if $@;
}
# ----------------------------------------------------------------
sub require_unicode_string {
    return if defined $Unicode::String::VERSION;
    local $@;
    eval { require Unicode::String; };
    die "Unicode::String module is required.\n" if $@;
}
# ----------------------------------------------------------------
sub require_unicode_map {
    return if defined $Unicode::Map::VERSION;
    local $@;
    eval { require Unicode::Map; };
    die "Unicode::Map module is required.\n" if $@;
}
# ----------------------------------------------------------------
1;
# ----------------------------------------------------------------


lib/Lingua/ZH/Romanize/Cantonese.pm
=head1 NAME

Lingua::ZH::Romanize::Cantonese - Romanization of Cantonese language

=head1 SYNOPSIS

    use Lingua::ZH::Romanize::Cantonese;

    my $conv = Lingua::ZH::Romanize::Cantonese->new();
    my $roman = $conv->char( $hanji );
    printf( "<ruby><rb>%s</rb><rt>%s</rt></ruby>", $hanji, $roman );

    my @array = $conv->string( $string );
    foreach my $pair ( @array ) {
        my( $raw, $ruby ) = @$pair;
        if ( defined $ruby ) {
            printf( "<ruby><rb>%s</rb><rt>%s</rt></ruby>", $raw, $ruby );
        } else {
            print $raw;
        }
    }

=head1 DESCRIPTION

This is Cantonese version of L<Lingua::ZH::Romanize::Pinyin> module.

Cantonese is one of the major dialects of Chinese language.
Its pronounciation is different from Mandarin Standard's.

=head1 DICTIONARY

This module's Hanji to roman mapping table is based on
both of CTLau.tit and CTLauBig5.tit which are distributed with cxterm.
Original files were prepared by Fung Fung Lee.

Sidney Lau's Cantonese transcription scheme as described in his book
"Elementary Cantonese", The Government Printer, Hong Kong, 1972.

=head1 SEE ALSO

L<Lingua::ZH::Romanize::Pinyin>

=head1 COPYRIGHT

Copyright (c) 2003-2006 Yusuke Kawasaki. All rights reserved.

=head1 LICENSE

Any commercial use of the Software requires a license directly from
the author(s).  Please contact the author(s) to negotiate an
appropriate license.  Commercial use includes integration of all or
part of the binary or source code covered by this permission
notices into a product for sale or license to third parties on your
behalf, or distribution of the binary or source code to third
parties that need it to utilize a product sold or licensed on your
behalf.

=cut
# ----------------------------------------------------------------
    package Lingua::ZH::Romanize::Cantonese;
    use strict;
    use vars qw( $VERSION @ISA );
    $VERSION = "0.12";
    @ISA = qw( Lingua::ZH::Romanize::Pinyin );
    use Lingua::ZH::Romanize::Pinyin;
# ----------------------------------------------------------------
#   every methods are provided by Lingua::ZH::Romanize::Pinyin!
# ----------------------------------------------------------------
;1;
# ----------------------------------------------------------------


lib/Lingua/ZH/Romanize/Pinyin.pm
=head1 NAME

Lingua::ZH::Romanize::Pinyin - Romanization of Standard Chinese language

=head1 SYNOPSIS

    use Lingua::ZH::Romanize::Pinyin;

    my $conv = Lingua::ZH::Romanize::Pinyin->new();
    my $roman = $conv->char( $hanji );
    printf( "<ruby><rb>%s</rb><rt>%s</rt></ruby>", $hanji, $roman );

    my @array = $conv->string( $string );
    foreach my $pair ( @array ) {
        my( $raw, $ruby ) = @$pair;
        if ( defined $ruby ) {
            printf( "<ruby><rb>%s</rb><rt>%s</rt></ruby>", $raw, $ruby );
        } else {
            print $raw;
        }
    }

=head1 DESCRIPTION

Pinyin is a phonemic notation of Chinese characters.

=head2 $conv = Lingua::ZH::Romanize::Pinyin->new();

This constructer methods returns a new object with its dictionary cached.

=head2 $roman = $conv->char( $hanji );

This method returns romanized letters of a Hanji character.
It returns undef when $hanji is not a valid Hanji character.
The argument's encoding must be UTF-8.
Both of Simplified Chinese and Traditional Chinese are allowed.

=head2 $roman = $conv->chars( $string );

This method returns romanized letters of Hanji characters.

=head2 @array = $conv->string( $string );

This method returns a array of referenced arrays
which are pairs of a Hanji chacater and its romanized letters.

    $array[0]           # first Chinese character's pair (array)
    $array[1][0]        # secound Chinese character itself
    $array[1][1]        # its romanized letters

=head1 DICTIONARY

This module's Hanji to roman mapping table is based on
PY.tit file which is distributed with cxterm.

=head1 MODULE DEPENDENCIES

L<Storable> module is required.

=head1 SEE ALSO

L<Lingua::ZH::Romanize::Cantonese>
L<Lingua::JA::Romanize::Japanese>
L<Lingua::KO::Romanize::Hangul>

http://www.kawa.net/works/perl/romanize/romanize-e.html

=head1 AUTHOR

Yusuke Kawasaki http://www.kawa.net/

=head1 COPYRIGHT

Copyright (c) 2003-2006 Yusuke Kawasaki. All rights reserved.

=head1 LICENSE

Any commercial use of the Software requires a license directly from
the author(s).  Please contact the author(s) to negotiate an
appropriate license.  Commercial use includes integration of all or
part of the binary or source code covered by this permission
notices into a product for sale or license to third parties on your
behalf, or distribution of the binary or source code to third
parties that need it to utilize a product sold or licensed on your
behalf.

=cut
# ----------------------------------------------------------------
    package Lingua::ZH::Romanize::Pinyin;
    use strict;
    use Carp;
    use Storable;
    use vars qw( $VERSION );
    $VERSION = "0.12";
# ----------------------------------------------------------------
sub new {
    my $package = shift;
    my $store = shift || &_detect_store( $package );
    Carp::croak "$! - $store\n" unless ( -r $store );
    my $self = Storable::retrieve( $store ) or Carp::croak "$! - $store\n";
    bless $self, $package;
    $self;
}
# ----------------------------------------------------------------
sub char {
    my $self = shift;
    my $char = shift;
    return unless exists $self->{$char};
    $self->{$char};
}
# ----------------------------------------------------------------
sub chars {
    my $self = shift;
    my @array = $self->string( shift );
    join( " ", map {$#$_>0 ? $_->[1] : $_->[0]} @array );
}
# ----------------------------------------------------------------
sub string {
    my $self = shift;
    my $src = shift;
    my $array = [];
    while ( $src =~ /([\300-\377][\200-\277]+)|([\000-\177]+)/sg ) {
        if ( defined $1 ) {                             # Chinese
            my $pair = [ $1 ];
            $pair->[1] = $self->{$1} if exists $self->{$1};
            push( @$array, $pair );
        } else {
            push( @$array, [ $2 ] );                    # ASCII
        }
    }
    @$array;
}
# ----------------------------------------------------------------
#   Pinyin.pm -> Pinyin.store
#   Cantonese.pm -> Cantonese.store
# ----------------------------------------------------------------
sub _detect_store {
    my $package = shift;
    my $store = $INC{join( "/", split("::","$package.pm"))};
    $store =~ s#\.pm$#.store# or Carp::croak "Invalid module name: $package\n";
    $store;
}
# ----------------------------------------------------------------
;1;
# ----------------------------------------------------------------
